File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        stack =[-1]
        res = 0
        n = len(heights)
        for i in range(n):
            while(stack[-1] != -1):
                top = heights[stack[-1]]
                if heights[i] < top:
                    popped = stack.pop()
                    height = heights[popped]
                    width = i - stack[-1]-1
                    res = max(res, height*width)
                else:
                    break
            stack.append(i)
                 #   height = 
        while(stack[-1] != -1):
            height = heights[stack.pop()]
            width = n - stack[-1] - 1
            res = max(res, height*width)
        return res
        
        

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        //leftLargest
        LinkedList<int[]> leftStack = new LinkedList<>();
        int N = heights.length;
        int[] leftSmallest = new int[N];
        int[] rightSmallest = new int[N];
        
        for(int i = 0; i < N; i++){
            int height = heights[i];
            rightSmallest[i] = N;
            //left largest
            while(!leftStack.isEmpty() && leftStack.peekLast()[0] >= height){
                int[] popped = leftStack.pollLast();
                rightSmallest[popped[1]] = i;
                }
                if(leftStack.isEmpty()){
                    leftSmallest[i] = -1;
                }else{
                    leftSmallest[i] = leftStack.peekLast()[1];
                }
            leftStack.addLast(new int[]{height, i});
        }
        int ans = 0;
        for(int j = 0; j < N; j++){
            int curHeight = heights[j];
            int left = leftSmallest[j], right = rightSmallest[j];
          //  System.out.printf("%d %d\n", left, right);
            if(left != j){
                left = left + 1;
            }
            if(right != j){
                right = right - 1;
            }
            ans = Math.max(ans, (right - left + 1) * curHeight);
        }
        return ans;      
    }
}